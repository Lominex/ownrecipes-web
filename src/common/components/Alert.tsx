import { Alert as BootstrapAlert } from 'react-bootstrap';

import '../css/alert.css';

import Icon from './Icon';
import P from './P';

export interface IAlertProps {
  severity: 'danger' | 'info';
  title: string;
  className?: string;

  children: React.ReactNode;
}

const Alert: React.FC<IAlertProps> = ({ severity, title, className, children }: IAlertProps) => (
  <BootstrapAlert variant={severity} className={className}>
    <BootstrapAlert.Heading>
      {severity === 'info'   && <Icon icon='info-circle' size='2x' className='alert-icon' />}
      {severity === 'danger' && <Icon icon='exclamation-diamond' size='2x' className='alert-icon' />}
      {title}
    </BootstrapAlert.Heading>
    {typeof children === 'string' && (
      <P className='alert-message mb-0'>{children}</P>
    )}
    {typeof children !== 'string' && (
      <div className='alert-message'>{children}</div>
    )}
  </BootstrapAlert>
);

export default Alert;
