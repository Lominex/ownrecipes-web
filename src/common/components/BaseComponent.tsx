import { Component } from 'react';
import { Form } from 'react-bootstrap';
import classNames from 'classnames';

import '../css/button.css';
import '../css/form_group.css';

import Icon from './Icon';

export interface IBaseComponentProps {
  name:       string;
  label?:     string;

  autoFocus?: boolean;
  required?:  boolean;
  readOnly?:  boolean;
  disabled?:  boolean;

  autoComplete?: string;
  helpText?:  React.ReactNode;
  tooltip?:   React.ReactNode;
  errors?:    React.ReactNode;

  className?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  change?: (name: string, newValue: any) => void;
}

export type BaseLabelProps = {
  htmlFor?: string;
}

export default class BaseComponent<P extends IBaseComponentProps, S = {}> extends Component<P, S> {
  constructor(props: P) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleChange(event: any) {
    if (this.props.readOnly || this.props.disabled) return;
    if (this.props.change) {
      this.props.change(this.props.name, event.target.value);
    }
  }

  hasError(): boolean {
    return !!this.props.errors;
  }

  getErrorMessage() { // eslint-disable-line react/no-unused-class-component-methods
    return this.hasError() ? <Form.Text className='error-text'>{this.props.errors}</Form.Text> : null;
  }

  getHelpText() { // eslint-disable-line react/no-unused-class-component-methods
    return this.props.helpText ? <Form.Text className='help-text'>{this.props.helpText}</Form.Text> : null;
  }

  getLabel(labelProps?: BaseLabelProps): React.ReactNode | undefined { // eslint-disable-line react/no-unused-class-component-methods
    if (this.props.label == null) {
      return null;
    } else if (this.props.tooltip) {
      return (
        <>
          <Form.Label htmlFor={labelProps?.htmlFor}>{this.props.label}</Form.Label>
          &nbsp;
          <Icon icon='info-circle' className='tooltip-icon' />
        </>
      );
    } else {
      return <Form.Label htmlFor={labelProps?.htmlFor}>{this.props.label}</Form.Label>;
    }
  }

  getFormGroupClassNames() { // eslint-disable-line react/no-unused-class-component-methods
    return classNames('form-group', this.props.className, {
      error:    this.hasError(),
      readonly: this.props.readOnly,
      required: this.props.required && !this.props.readOnly,
      'no-label': this.props.label == null,
    });
  }
}
