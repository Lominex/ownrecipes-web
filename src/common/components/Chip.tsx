import classNames from 'classnames';

import '../css/chip.css';

export interface IChipProps {
  variant: 'primary' | 'secondary';

  className?: string;
  children: React.ReactNode;
}

const Chip: React.FC<IChipProps> = (props: IChipProps) => (
  <span
      className={classNames('chip', props.variant)}>
    {props.children}
  </span>
);

export default Chip;
