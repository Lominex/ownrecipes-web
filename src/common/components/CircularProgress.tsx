/** Code adapted from https://github.com/KyleAMathews/react-spinkit */

import classNames from 'classnames';

import '../css/circular_progress.css';

export interface ICircularSpinnerProps {
  /**
   * Variant of the spinner.
   *
   * @defaultValue `circle`
   */
  variant?: 'circle' | 'three-bounce';
  /**
   * Color of the spinner. When `undefined` the color will be neutral.
   */
  color?: 'primary' | 'secondary';
  /** Custom className for styling. */
  className?: string;
}

function getDivCount(variant: 'circle' | 'three-bounce'): number {
  switch (variant) {
    case 'circle':       return 12;
    case 'three-bounce': return  3;
    default: throw new Error(`Invalid argument: The spinner variant "${variant}" is invalid.`);
  }
}

const CircularProgress: React.FC<ICircularSpinnerProps> = ({
    variant, color, className }: ICircularSpinnerProps) => {
  const variantD = variant ?? 'circle';

  const classes = classNames('circular-progress', variantD, color, className);
  const divCount = getDivCount(variantD);

  return (
    <div className={classes}>
      {/* eslint-disable-next-line react/no-array-index-key */}
      {[...Array(divCount)].map((_, idx) => <div key={idx} />)}
    </div>
  );
};

export default CircularProgress;
