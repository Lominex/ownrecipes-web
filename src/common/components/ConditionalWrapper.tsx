interface IConditionalWrapperProps {
  condition: boolean;
  render: (children: any) => React.ReactElement; // eslint-disable-line @typescript-eslint/no-explicit-any
  children: React.ReactNode;
}

const ConditionalWrapper: React.FC<IConditionalWrapperProps> = ({ condition, render, children }: IConditionalWrapperProps) => {
  if (children == null) return null;
  return condition ? render(children) : <>{children}</>; // eslint-disable-line react/jsx-no-useless-fragment
};

export default ConditionalWrapper;
