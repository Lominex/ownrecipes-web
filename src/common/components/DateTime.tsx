import Datetime from 'react-datetime';
import moment from 'moment';
import { BaseComponent, IBaseComponentProps, IBaseComponentState } from './FormComponents';

require('react-datetime/css/react-datetime.css');

interface IDateTimeProps extends IBaseComponentProps {
  id: string;
  name: string;
  label: string;
  timeFormat?: string;
  dateFormat?: string;
  change: (name: string, newIsoDate: string) => void;
  class: string;
}

interface IDateTimeState extends IBaseComponentState {
  value: moment.MomentInput,
}

export default class DateTime extends BaseComponent<IDateTimeProps, IDateTimeState> {
  handleChange(date: moment.MomentInput) {
    this.setState({
      value: date,
    });

    if (this.props.change) {
      this.props.change(this.props.name, moment(date).toISOString());
    }
  }

  render() {
    return (
      <div className={this.props.class} key={this.props.id}>
        <div className={`form-group ${this.hasErrors() ? 'has-error' : null}`}>
          { this.props.label ? <label>{this.props.label}</label> : null }
          <Datetime
              inputProps={{ name: this.props.name, className: 'form-control' }}
              onChange={this.handleChange}
              value={!this.props.timeFormat ? moment(this.state.value).format('ddd, ll') : moment(this.state.value).format('llll')}
              dateFormat={this.props.dateFormat || 'ddd, ll'}
              timeFormat={this.props.timeFormat} />
          {this.getErrorMessage()}
        </div>
      </div>
    );
  }
}
