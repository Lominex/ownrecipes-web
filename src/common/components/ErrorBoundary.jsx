import { Component } from 'react';

const ErrorBoundaryContainer = props => (
  <div>
    {props.children}
  </div>
);

/**
 * Simple error catcher.
 *
 * Wrap the usage of components to prevent any error from leaving.
 *
 * Example:
 * ```
 * <ErrorBoundary>
 *     <MyComponent ... />
 * </ErrorBoundary>
 * ```
 *
 * See also: {@link https://reactjs.org/docs/error-boundaries.html}
 */
class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hasError: false,
      error: undefined,
    };
  }

  static getDerivedStateFromError(error) {
    return { error: error, hasError: true };
  }

  componentDidCatch(error: Error | null) {
    const { t } = this.props;
    this.setState({ error: error || new Error(t('error:missing error')) });
  }

  render() {
    if (this.state.hasError && this.state.error) {
      return (
        <ErrorBoundaryContainer>
          <h1>Error</h1>
          {this.props.verbose && (
            <>
              <h2>Message:</h2>
              <p>{this.state.error.message}</p>
            </>
          )}
          {this.props.printStack && (
            <>
              <h2>Stack:</h2>
              <p>{this.state.error.stack}</p>
            </>
          )}
        </ErrorBoundaryContainer>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
