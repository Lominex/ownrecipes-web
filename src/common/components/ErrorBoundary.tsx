import { Component } from 'react';

import { AnyComponent } from '../../types/Types';
import P from './P';

interface IErrorBoundaryProps {
  /** Print message? */
  verbose:    boolean;
  /** Print error stack? */
  printStack: boolean;
  /** The error. */
  children: AnyComponent;
}

interface IErrorBoundaryState {
  readonly hasError: boolean;
  readonly error: Error | null | undefined;
}

interface IErrorBoundaryContainerProps {
  /** The error. */
  children: AnyComponent;
}

const ErrorBoundaryContainer: React.FC<IErrorBoundaryContainerProps> = ({ children }: IErrorBoundaryContainerProps) => (
  <div style={{ margin: '0.5rem' }}>
    {children}
  </div>
);

/**
 * Simple error catcher.
 *
 * Wrap the usage of components to prevent any error from leaving.
 *
 * Example:
 * ```
 * <ErrorBoundary>
 *     <MyComponent ... />
 * </ErrorBoundary>
 * ```
 *
 * See also: {@link https://reactjs.org/docs/error-boundaries.html}
 */
 class ErrorBoundary extends Component<IErrorBoundaryProps, IErrorBoundaryState> {
  constructor(props: IErrorBoundaryProps) {
    super(props);
    this.state = {
      hasError: false,
      error: undefined,
    };
  }

  static getDerivedStateFromError(error: Error): IErrorBoundaryState {
    return { error: error, hasError: true };
  }

  componentDidCatch(error: Error | null) {
    this.setState({ error: error || new Error('Error missing') });
  }

  render() {
    if (this.state.hasError && this.state.error) {
      return (
        <ErrorBoundaryContainer>
          <h1>Error</h1>
          {this.props.verbose && (
            <>
              <h2>Message:</h2>
              <P>{this.state.error.message}</P>
            </>
          )}
          {this.props.printStack && (
            <>
              <h2>Stack:</h2>
              <P>{this.state.error.stack}</P>
            </>
          )}
        </ErrorBoundaryContainer>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
