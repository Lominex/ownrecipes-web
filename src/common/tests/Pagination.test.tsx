import create from '../../test/create';
import { PaginationLink } from '../components/Pagination';

test('PaginationLink', () => {
  const buildUrl = (): string => 'http://localhost:8080';
  const component = create(
    <PaginationLink
        title='17'
        offset={128}
        active
        buildUrl={buildUrl}
        />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('PaginationLink disabled', () => {
  const buildUrl = (): string => 'http://localhost:8080';
  const component = create(
    <PaginationLink
        title='<-'
        offset={128}
        disabled
        buildUrl={buildUrl}
        />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
