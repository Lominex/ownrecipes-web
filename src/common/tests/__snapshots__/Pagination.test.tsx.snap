// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationLink 1`] = `
<li
  className="page-item active"
>
  <a
    className="page-link btn-outline-primary active"
    href="/http://localhost:8080"
    onClick={[Function]}
  >
    17
  </a>
</li>
`;

exports[`PaginationLink disabled 1`] = `
<li
  className="page-item disabled"
>
  <a
    className="page-link btn-outline-primary disabled"
    href="/"
    onClick={[Function]}
  >
    &lt;-
  </a>
</li>
`;
